relays/finality-relay/src/finality_loop_tests.rs

```diff
@@ -0,0 +1,366 @@
+	let client_data = run_sync_loop(|data| {
+		// header#7 has persistent finality proof, but it isn't mandatory => it isn't submitted, because
+		// header#8 has persistent finality proof && it is mandatory => it is submitted
+		// header#9 has persistent finality proof, but it isn't mandatory => it is submitted, because
+		//   there are no more persistent finality proofs
+		//
+		// once this ^^^ is done, we generate more blocks && read proof for blocks 12, 14 and 16 from the stream
+		// but we only submit proof for 16
+		//
+		// proof for block 15 is ignored - we haven't managed to decode it
+		if data.target_best_block_number == 9 {
+			data.source_best_block_number = 17;
+			data.source_headers.insert(11, (TestSourceHeader(false, 12), None));
+			data.source_headers
+				.insert(12, (TestSourceHeader(false, 12), Some(TestFinalityProof(Some(12)))));
+			data.source_headers.insert(13, (TestSourceHeader(false, 13), None));
+			data.source_headers
+				.insert(14, (TestSourceHeader(false, 14), Some(TestFinalityProof(Some(14)))));
+			data.source_headers
+				.insert(15, (TestSourceHeader(false, 15), Some(TestFinalityProof(None))));
+			data.source_headers
+				.insert(16, (TestSourceHeader(false, 16), Some(TestFinalityProof(Some(16)))));
+			data.source_headers.insert(17, (TestSourceHeader(false, 17), None));
+		}
+
+		data.target_best_block_number == 16
+	});
+
+	assert_eq!(
+		client_data.target_headers,
+		vec![
+			(TestSourceHeader(true, 8), TestFinalityProof(Some(8))),
+			(TestSourceHeader(false, 9), TestFinalityProof(Some(9))),
+			(TestSourceHeader(false, 16), TestFinalityProof(Some(16))),
+		],
+	);
+}
+
+#[test]
+fn prune_unjustified_headers_works() {
+	let mut original_unjustified_headers: UnjustifiedHeaders<TestFinalitySyncPipeline> = vec![
+		TestSourceHeader(false, 10),
+		TestSourceHeader(false, 13),
+		TestSourceHeader(false, 15),
+		TestSourceHeader(false, 17),
+		TestSourceHeader(false, 19),
+	]
+	.into_iter()
+	.collect();
+
+	// when header is in the collection
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(10, &mut unjustified_headers),
+		Some(TestSourceHeader(false, 10)),
+	);
+	assert_eq!(
+		&original_unjustified_headers.clone().make_contiguous()[1..],
+		unjustified_headers.make_contiguous(),
+	);
+
+	// when there's no header in the collection
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(11, &mut unjustified_headers),
+		None,
+	);
+	assert_eq!(
+		&original_unjustified_headers.clone().make_contiguous()[1..],
+		unjustified_headers.make_contiguous(),
+	);
+
+	// when last entry is pruned
+	let mut unjustified_headers = original_unjustified_headers.clone();
+	assert_eq!(
+		prune_unjustified_headers::<TestFinalitySyncPipeline>(19, &mut unjustified_headers),
+		Some(TestSourceHeader(false, 19)),
+	);
```

Comment:
Is there anything special about the last entry? Just trying to figure out why you have a test for it explicitly

Reply:
The implementation uses `binary_search_by_key` and `split_off` - there's a lot of ways you could screw up with last element when using this function